### ISSUE TITLE
Input not being recognized for N64 using front end.

### ISSUE THREAD

**Issue by stuffu** [libretro/RetroArch#15109]
Mupen64 does not record input when using front end. It does for some games strangely but the majority it doesn’t.

I don’t think this only applies to mupen64,
Also tested other N64 results with the same result, but mupen64 is the one I would like to use.

Launching a game from within RA, all key presses are fine.

I’m using daijisho on front end on android in this case but read other posts about people having the same issue with other front ends as well, also with other os.

All other systems works as expected, just N64 that behaves this way and according to issue on daijisho, it’s not related to the front end itself, but a bug in RA.

**Comment by hizzlekizzle** [comment]
Do you have a link to the daijisho issue discussing this? it would be nice if they had any hints about what could be going wrong, since it works fine from RetroArch itself.

@jasonunbrokensoftware Do you have this issue on your end?

**Comment by stuffu** [comment]
Here is the link to the issue, but I have seen it with other frontends as well.

https://github.com/magneticchen/Daijishou/issues/181

**Comment by jasonunbrokensoftware** [comment]
It doesn't look like there are any issues with controller input when launching games from LaunchBox through the Parallel N64 core. I've been using that for quite a long time, and I just tested it and everything is working fine. So I don't really understand the issue. I know if input was broken for the N64 cores through LaunchBox we'd have tons of reports of issues. Regardless, for what it's worth I've been testing via Bluetooth with an Xbox Series X controller.

**Comment by stuffu** [comment]
Can I contribute with something to narrow it down?

**Comment by stuffu** [comment]
> I guess if you could pinpoint which cores exhibit this behavior, we can see if there's a common thread among them.

I have tried mupen64_next_gles2/3 and parallele64 same behavior with either core.

**Comment by stuffu** [comment]
> What specific Android devices are affected? 

In my case Retroid Pocket 3+, which is great for handheld gaming. I don’t see any specific devices mentioned in other posts so not sure how widespread it is.

>What version of Android are the affected devices running? 

Again, in my case Android 11.

>What specific controllers have been tested? Do any controllers not have this issue? >Have both Bluetooth and USB controllers been tested?

These devices have built in controls, think Nintendo switch but for retro games. The strange part here is that all other cores work fine, even running dolphin, scummvm point and click, amiga games with mouse input etc.

**Comment by stuffu** [comment]
I should also mentioned that tried to launch a few games. 007 and super Mario worked as expected, resident evil 2 did not recognize any inputs. 

The majority of the games does not recognize input. For reference, I previously had RA on Xbox S with dev mode without any issues and with the same game collection.

**Comment by jasonunbrokensoftware** [comment]
Okay, I think it has to be assumed that it's an issue with that specific device, at least until someone else comes forward that has run into the same thing on a different device. Unfortunately I don't have any means of testing that without going out and buying that device, so it may be difficult to figure out.

**Comment by unbelievableflavour** [comment]
Hi, as show in the issues above I have the exact same problem. Here is some extra information that might help with the issues.

**Steps to reproduce:**
Launch an instance of mupen using the following config::
```
-n com.retroarch.aarch64/com.retroarch.browser.retroactivity.RetroActivityFuture
-e ROM /PATH/TO/FILE
-e LIBRETRO /data/data/com.retroarch.aarch64/cores/mupen64plus_next_gles3_libretro_android.so
-e CONFIGFILE /storage/emulated/0/Android/data/com.retroarch.aarch64/files/retroarch.cfg\n  
-e IME com.android.inputmethod.latin/.LatinIME
-e DATADIR /data/data/com.retroarch.aarch64
-e APK /data/app/com.retroarch.aarch64-1/base.apk
-e SDCARD /storage/emulated/0
-e EXTERNAL /storage/emulated/0/Android/data/com.retroarch.aarch64/files
--activity-clear-task
--activity-clear-top
```


**Environment information**
OS: Android 11, retroid pocket 3+

**Important**
This issue does not happen with every game.

Try one of the following
- Harvest Moon 64(USA)
- Star Wars: Rogue Squadrons(USA)
- Star Wars: Episode 1 Racer(USA)
- Super Smash Bros(USA)
- Tonic Trouble(USA)
- Rakugakids(EU)

**Comment by jasonunbrokensoftware** [comment]
Okay, well then assuming that all of @unbelievableflavour 's testing is correct (thank you @unbelievableflavour for all the detailed information), that would mean that the issue is likely specific to Retroid Pocket devices, which means that it's most likely a bug with the Retroid firmware/Android installation. It might be wise to report it to Retroid themselves, as they're the most likely to be able to fix the bug.

There may be a workaround that could be implemented by the RetroArch team, but whoever works on this will definitely need a Retroid Pocket 3+ device on hand, and they may or may not be able to do anything about it, depending on how bad the bug in the firmware actually is.

If anyone finds any other devices that are affected (other than Retroid devices), that may suggest that there is an actual bug in RetroArch and not just the Retroid firmware. But until then, it should probably be assumed that it's a Retroid firmware issue.

**Comment by stuffu** [comment]
Thanks, I have asked retroid on their discord to look into it.

**Comment by unbelievableflavour** [comment]
Thanks. Im also still trying to find someone who will test on a device with physical buttons other than retroid. I think we still gotta rule that out.

**Comment by stuffu** [comment]
I got a reply with a workaround and that the issue has been forwarded to the engineers. The workaround is to do button presses immediately after launching a game. I tried with Resident Evil 2, it works!

**Comment by unbelievableflavour** [comment]
@stuffu yes. I've listed that workaround here too:
https://github.com/libretro/mupen64plus-libretro-nx/issues/485
https://github.com/libretro/parallel-n64/issues/758

What else did the developers say? And by developers you mean Retroid? Will they look into the issue?

**Comment by stuffu** [comment]
@unbelievableflavour Yes, they said they will look into it, but not sure what that really means...

**Comment by unbelievableflavour** [comment]
Thanks. I also got a reply from Retroid support: Their exact reply: 

"Hi, I have reported this issue to our engineers and they will try their best to improve this issue, maybe the next OTA will fix this issue.
There are two temporary solutions to this problem. The first is to quickly press the button after starting the game. The second is to enter the menu through the virtual button of RA after starting the game, and then press the button.
Thank you for your support of our products."

**Comment by cbrumbau** [comment]
What specific Android devices are affected? Odin Pro
What version of Android are the affected devices running? 10
What specific controllers have been tested? The built in controller

As such it's more than just Retroid Pocket devices affected by this issue

Just did a quick test using Daijishou and found Kirby 64 had the input issue while Mario Kart 64 was fine

**Comment by unbelievableflavour** [comment]
@jasonunbrokensoftware Since Odin Pro also has this issue now. It is no longer a Retroid specific issue.

**Comment by jasonunbrokensoftware** [comment]
Good to know. Unfortunately I don't have either of those devices so I'm not likely to be much help here, but hopefully that will help to narrow it down.

**Comment by delta62** [comment]
Hi all, reporting in with another handheld that has similar symptoms:

| Parameter | What I tested |
| -------------- | ------------- |
| Device       | Odin Lite |
| Core          | Mupen64Plus-Next GLES3, ParaLLEl N64 |
| Frontend  | Daijishou 1.4.62 (394), LaunchBox 1.9 |
| Android version | 11 |
| Controller           |  Built-in gamepad, 8BitDo SN30 Pro, Xbox One Controller |

In my case, the posted workarounds of either pressing buttons immediately after launching or using the touch screen do not restore functionality of any hardware controls. I _can_ control the game using the on-screen controls, but the built in buttons & controllers only work if I start the game via RA directly.

To be clear, I tested with every combination of frontend + controller above and always had the same results. In my case the game did not appear to make any difference, it never worked. The only way controllers of any kind worked for me were from launching the game within RetroArch. This worked 100% of the time.

I also tested configuring Daijishou and LaunchBox to launch Mupen64 FZ rather than RetroArch, and things work correctly then.

Also, this interestingly only effects the N64 cores as far as I can tell. I also tested SNES and GBA and both worked fine with each frontend & hardware controller, with the frontend launching RA.

**Comment by loki666** [comment]
I had the same issue with my Backbone One.
I changed input polling to Early.

No issues anymore.

**Comment by edbanshee** [comment]
I'm having the same problem. I'm on Retroarch Nightly 1.16.0
Device: Poco X3 Pro, Android 13.
My gamepad is BSP-3, tried different input modes and none of them work. 
Daijisho Frontend
Tried with Parallel64 and MupenGLES3 cores. I'm only having issues with those, every other core I have works fine (this includes many other consoles like Dreamcast, SNES, GB, GameCube, Wii, etc)
Diddy Kong Racing worked, the majority of the other games did not.

**Comment by unbelievableflavour** [comment]
@loki666 fix worked for me! Thank you.

Change this setting like so:

1. Go to Retroarch > Settings > Input > Polling Behaviour
2. Change to "Early"
3. Restart retroarch.

**Comment by feress737** [comment]
I think all Games that support rumble are not working from a frontend without "Polling behavior = early". With set to "early" the input works, but no rumble. Both works if I start a game directly over retroarch.

**Comment by Kevinwensh93** [comment]
> [unbelievableflavour](/unbelievableflavour)

Want to add that this also happens to Odin Mini 2 with ES-DE Frontend.

Setting polling = early will fix the issue of recognizing the controller but rumble will never be operational.

**Comment by lavosprime** [comment]
I don't think this issue is specific to N64. I have been having the same problem (no input, even hotkeys) with the same solution (early polling, or open game from within Retroarch) in SNES games on a Retroid Pocket Mini running Android 10, using the ES-DE frontend and the bsnes-hd core. (I like 1:1 PAR but still want to use the whole 4:3 screen...)

I have access to a GameMT E5+, and I would be willing to A/B test on that device sometime, but I don't currently have the free time at a PC to go through all the necessary setup.

**Comment by vifon** [comment]
I can confirm the issue and the polling=early fix on the following setup:

- Device: Ayn Thor
- OS: Android 13
- Core: Mupen64Plus-Next GLES3
- Game: Kirby 64 - The Crystal Shards
- RetroArch version: 1.22.2

**Comment by SirPrimalform** [comment]
Just to add, I'm having the exact same issue on the Anbernic RG476H, using ES-DE as a launcher. This is definitely an issue with RA or the Mupen core rather than a specific launcher. Polling=early does help, but still no rumble.

**Comment by matheuswillder** [comment]
This used to happen on my cheap/low-end Android 11, using RetroArch with Mupen64Plus-Next and playing with the on-screen overlay. There are some issues reporting this in the Mupen64Plus-Next repository, but apparently the maintainer couldn't fix it on the core side. Links here: https://github.com/libretro/mupen64plus-libretro-nx/issues/57, and here: https://github.com/libretro/mupen64plus-libretro-nx/issues/317.

It don't happens on my current Android device, but there's a post on the RetroArch subreddit recently [with this issue as well](https://www.reddit.com/r/RetroArch/comments/1rgb4qq/oot_and_mm_no_controls/), so I decided to point it out here.

In all these cases, changing **Settings>Input>Polling Behavior** to **Early** fixes it.
